src/internal/pjs/api_server.go

```diff
@@ -0,0 +1,110 @@
+package pjs
+
+import (
+	"context"
+
+	"github.com/pachyderm/pachyderm/v2/src/internal/dbutil"
+	"github.com/pachyderm/pachyderm/v2/src/internal/errors"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pachsql"
+	"github.com/pachyderm/pachyderm/v2/src/internal/pjsdb"
+	"github.com/pachyderm/pachyderm/v2/src/internal/storage/fileset"
+	txnenv "github.com/pachyderm/pachyderm/v2/src/internal/transactionenv"
+	"github.com/pachyderm/pachyderm/v2/src/pjs"
+	"google.golang.org/grpc/codes"
+	"google.golang.org/grpc/status"
+)
+
+type apiServer struct {
+	pjs.UnimplementedAPIServer
+	env    Env
+	txnEnv *txnenv.TransactionEnv
+}
+
+func newAPIServer(env Env) *apiServer {
+	return &apiServer{
+		env:    env,
+		txnEnv: env.TxnEnv,
+	}
+}
+
+func (a *apiServer) CreateJob(ctx context.Context, request *pjs.CreateJobRequest) (response *pjs.CreateJobResponse, retErr error) {
+	var ret pjs.CreateJobResponse
+	if request.Input == nil {
+		return nil, status.Errorf(codes.InvalidArgument, "missing data input")
+	}
+	if err := dbutil.WithTx(ctx, a.env.DB, func(ctx context.Context, sqlTx *pachsql.Tx) error {
+		program, err := fileset.ParseID(request.Program)
```

Comment:
This stuff doesn't need to be in the transaction; you can move it above so the transaction is just the operation you need to perform.